ENH: use `deprecated`-directive in docstring, add `message_location` …

kmuehlbauer opened the pull request:
…to insert deprecation note at top or bottom of docstring

- This uses the `deprecation`-directive to add the deprecation note to the docstring.

- Via `message_location` you can choose to insert the deprecation note at the top (directly after the summary line, or at the bottom). Defaults to bottom.

kmuehlbauer commented:
@briancurtin Ah I see, I'll have to fix the tests. Also I did test this only on python 3.6 but `split` should have `maxsplit` also on python 2.7. 

Sure we can get this running.

briancurtin commented:
I think if you want to support some location configurability that it should probably be a module-level global instead of a parameter on the decorator. For example, I actually like the "top" option for something I used this decorator for recently, but I would have to type `location="top"` many times since I'd want the several uses in the project to look the same (and I imagine people are going to want this to be consistent).

Having the location option in the decorator's parameter list also makes it read slightly off to me—the version numbers and details are easily understood by someone reading the code, but the location parameter doesn't feel like a natural fit there. You'd see "oh it's deprecated in 1.0 and will be removed in 2.0, and I should use the bar function instead, that's great...but hmm, what's location=top?"

What if you had `deprecation.message_location = "bottom"` and then all uses of `deprecation.deprecated` use that?

kmuehlbauer commented:
@briancurtin Makes absolutely sense. I'll update my PR tomorrow.

kmuehlbauer commented:
@briancurtin I updated my PR according to your suggestions regarding `message_location`. 

I fixed the tests to work with the `deprecation`-directive and added a test to check for correct `top`-location.

Unfortunately I do not see immediately why the coverage went down.

kmuehlbauer commented:
@briancurtin 

Sorry for the possible confusion. After re-reading your statement I thought I didn't made all changes per your suggestion. I now completely removed the keyword argument. This means that `deprecation.message_location` defaults to `bottom`. If someone wants to insert the location to top, `deprecation.message_location` has to be changed once per source file (IIRC).

kmuehlbauer commented:
OK, it seems that `keep it simple` is best in this case. Now this falls silently back to `bottom`-location if message_location is not `top`. 

@briancurtin I think, this is good to go from my end, finally.

briancurtin commented on deprecation.py:
```diff
@@ -157,20 +160,23 @@ def _function_wrapper(function):
-                    ", to be removed in %s" % removed_in if removed_in else "",
-                "period":
-                    "." if deprecated_in or removed_in or details else "",
+                    "\n   To be removed in %s." %
```
Now that the deprecated and removed parts are in two different sentences, what about changing "To be removed in" to "This will be removed in"? It read ok before when it was one sentence split by a comma, but I think it could be better now that it's two sentences.

kmuehlbauer replied:
Just a second!

kmuehlbauer commented:
@briancurtin It would be perfect, if you could make this a release also.
